### DISCUSSION TITLE
IKE_SA is being deleted after expiring inactivity time even though there is an activity

### DISCUSSION THREAD

**Discussion by sridharkondeti** [strongswan/strongswan#1875] [General]
Hi Team,

I have established strongswan connection between 2 VMs. 
I have setup one VM as client and one VM as server.
On server strongswan config I added “inactivity=1800”.

I mounted an NFS file system from server to the client and continuously read/write data(with fio command) to this mounted file system to keep the connection active.

But the strongswan on the server sent delete request saying inactivity time elapsed even though there was continuous data flow between the servers.
“deleting IKE_SA after 1800 seconds of CHILD_SA inactivity”

Can anybody help me in understanding why this inactivity action is being triggered?

Thanks!

**Comment by tobiasbrunner** [comment]
> Can anybody help me in understanding why this inactivity action is being triggered?

Maybe the traffic does not go via VPN or the stats are not correctly updated. Check the status output of strongSwan to see usage stats and maybe even via `ip -s xfrm state` and `ip -s xfrm policy` directly in the kernel.

**Comment by sridharkondeti** [comment]
Thanks for response @tobiasbrunner.
The “ip -s xfrm state” command shows that the packets are being sent and received.
```
src x.x.x.1 dst x.x.x.2
	…
	lifetime current:
	  437752(bytes), 2555(packets)
	stats:
	  replay-window 0 replay 0 failed 0
src x.x.x.2 dst x.x.x.1
	…
	lifetime current:
	  415876(bytes), 1547(packets)
	stats:
	  replay-window 0 replay 0 failed 0
```

But “ip -x xfrm policy” command is showing as 0 packets/bytes.

I also checked the tcpdump, it shows that the packets are going out and coming in on port 4500 with UDP-encap.
```
20:15:04.788666 ens3  Out ifindex 2 02:00:0b:00:23:8c ethertype IPv4 (0x0800), length 372: x.x.x.1.4500 > x.x.x.2.4500: UDP-encap: ESP(spi=xx,seq=xx), length 324
20:15:04.789176 ens3  In  ifindex 2 02:00:07:00:23:8c ethertype IPv4 (0x0800), length 464: x.x.x.2.4500 > x.x.x.1.4500: UDP-encap: ESP(spi=xx,seq=xx), length 416
```
But still receiving the DELETE request saying due to inactivity.

**Reply by tobiasbrunner** [reply]
> But “ip -x xfrm policy” command is showing as 0 packets/bytes.

I guess `-x` is a typo (should be `-s`), but it's normal that the packet/bytes counters are not increasing for the policies, check the timestamps there, in particular `use`. It should change over time and with traffic (for both in/out and possibly fwd). Also, make sure to run this on the host that sends the delete.

**Reply by sridharkondeti** [reply]
Hi @tobiasbrunner,
I actually ran `ip -s xfrm policy` command only. It was a typo in the above comment.
I checked the traffic flow on the other end instance as well, it is also sending/receiving the packets.
Does having NAT affects this behaviour?

**Reply by tobiasbrunner** [reply]
> I actually ran ip -s xfrm policy command only. It was a typo in the above comment.
I checked the traffic flow on the other end instance as well, it is also sending/receiving the packets.

Then you have all the information to debug this yourself (e.g. check the use times before inactivity event occurs).

> Does having NAT affects this behaviour?

No, the inactivity timer is based on the last use time of the policies (or on newer kernels the SAs) of a particular CHILD_SA. So that's unrelated to UDP encapsulation.

**Reply by sridharkondeti** [reply]
Hi @tobiasbrunner ,
Sorry for the delay. I observed `ip -s xfrm state`, I see that `use` times are not being updated even though there is continuous traffic flowing through the IPSec connection.
I verified with `tcpdump` that the traffic is flowing through port 4500 and UDP encapsulated. I also see that the packet sizes are being updated with `ip -s xfrm state`.
Could you please suggest why the `use` times are not being updated, any config option should be updated?

**Reply by tobiasbrunner** [reply]
As I wrote before, the `use` time in `ip -s xfrm state` is **not** updated as it doesn't signify the "last" use time but the "first" use time (only recent kernels can provide the most recent use time on states). You have to look at the `use` time  in `ip -s xfrm policy` as that's updated for every packet that matches the policy.

**Reply by sridharkondeti** [reply]
Thanks @tobiasbrunner. Yes, I saw that `ip -s xfrm policy` showing the use times updated for every packet that matches the policy. But looks like the strongswan is sending `deleting IKE_SA after 1800 seconds of CHILD_SA inactivity` based on the `ip -s xfrm state` use times.

**Reply by tobiasbrunner** [reply]
> But looks like the strongswan is sending `deleting IKE_SA after 1800 seconds of CHILD_SA inactivity` based on the `ip -s xfrm state` use times.

What makes you think so? As I said before, the kernel does not update that use time, so strongSwan **never** uses it. Only on kernels >= 6.2, and with strongSwan 5.9.10 and newer, does the last use time get queried from the SAs (using a separate attribute, not the use time you're referring to), in all other cases the policies are queried (you should see that reflected in the stats when querying the SAs with `swanctl --list-sas`). If that doesn't work as expected, there might be some other issue on your system (e.g. with monotonic timestamps).

**Reply by sridharkondeti** [reply]
As that was the only time(`use` time from `ip -s xfrm state`) not getting updated, I thought that's being used to calculate `inactivity timeout`, but looks like that's not the case, thanks for clarification.
Yes, the stats are always being updated.
swanctl --list-sas` shows the updated stats.
  ```
 installed 2221s ago, rekeying in 1080s, expires in 1739s
    in  <spi>, 57856707926 bytes, 6505059 packets,     1s ago
   out <spi>, 57497392 bytes, 204561 packets,    32s ago
```
So something else is causing this inactivity timeout to happen, could be monotonic timestamps then.

**Reply by tobiasbrunner** [reply]
> could be monotonic timestamps then

Hm, since the time in the status output is apparently correctly updated, I don't think so. The `... ago` part uses the same calculation as the inactivity timeout: `time_monotonic(NULL) - <reported last use time>`. The inactivity job just uses the maximum of both directions and also the install time (see [here](https://github.com/strongswan/strongswan/blob/6f0cd19fd6a9feeb6872057723a8fc4bee884ceb/src/libcharon/processing/jobs/inactivity_job.c#L79-L83)) to avoid deleting newly created/rekeyed CHILD_SAs that have not yet been used.

Also, the log message you quoted in your original post indicates that you enabled the `charon.inactivity_close_ike` option, which isn't the default. Is that the case?

**Reply by sridharkondeti** [reply]
Yes, we enabled this option `charon.inactivity_close_ike` on remote host. We are still working on figuring out why DELETE requests are being received. Thanks for your help, @tobiasbrunner.
